File: clojure/src/project_euler/sol/p0026.clj
(ns project-euler.sol.p0026
  (:require
   [project-euler.lib.math :as math]
   [project-euler.lib.math.prime :as prime]))

(defn- pp
  "Remove prime factors 2 and 5 from `n`."
  ^long [n]
  (letfn [(aux [n denom]
            (loop [n n]
              (if (zero? (mod n denom))
                (recur (quot n denom))
                n)))]
    (aux (aux n 2) 5)))

;;; This function is not strictly the correct Carmichael function
;;; because the function assumes that the argument is not a multiple of 2.
(defn- carmichael
  [n]
  (->> (prime/factorize n)
       (map (fn [[base exp]] (* (dec base) (long (math/pow base (dec exp))))))
       (reduce math/lcm)))

(defn- find-repetend-length
  ^long [d]
  (let [d (pp d)]
    (if (= d 1)
      0
      (first (for [k (prime/divisors (carmichael d))
                   :when (= (math/powermod 10 k d) 1)]
               k)))))

(defn solve
  ([]
   (solve 1000))
  ([upper]
   (loop [nums (range (dec upper) (dec (quot upper 2)) -1)
          max-length 0
          answer 0]
     (when-first [i nums]
       (if (<= i max-length)
         answer
         (let [repetend-length (find-repetend-length i)]
           (if (> repetend-length max-length)
             (recur (rest nums) repetend-length (pp i))
             (recur (rest nums) max-length answer))))))))

File: rust/src/bin/p0026.rs
// Project Euler: Problem 26

euler::run_solver!(26);

fn solve() -> String {
    compute(1_000).to_string()
}

fn compute(upper: u64) -> u64 {
    debug_assert!(upper > 1);

    let mut max_length = 0_u64;
    let mut ans = 0_u64;

    for d in ((upper / 2)..upper).rev() {
        if d <= max_length {
            break;
        }
        if let Some(tpl @ (length, _)) = find_repetend_length(d)
            && length > max_length
        {
            (max_length, ans) = tpl;
        }
    }

    ans
}

// preprocessing
fn pp(mut n: u64) -> u64 {
    while n.is_multiple_of(2) {
        n /= 2;
    }
    while n.is_multiple_of(5) {
        n /= 5;
    }

    n
}

fn find_repetend_length(mut n: u64) -> Option<(u64, u64)> {
    use euler::math;

    // This function is not strictly correct Carmichael function
    // because the function assumes that the argument is not a multiple of 2.
    fn carmichael(n: u64) -> u64 {
        math::factorize(n)
            .into_iter()
            .map(|(b, e)| (b - 1) * b.pow(e - 1))
            .reduce(math::lcm)
            .unwrap()
    }

    n = pp(n);
    if n == 1 {
        return None;
    }
    for k in math::divisors(carmichael(n)) {
        if math::powmod(10, k, n) == 1 {
            return Some((k, n));
        }
    }

    unreachable!();
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0026_upper_10() {
        assert_eq!(compute(10), 7);
    }

    #[test]
    fn p0026_upper_300() {
        assert_eq!(compute(300), 289);
    }

    #[test]
    fn p0026_upper_1000() {
        assert_eq!(compute(1_000), 983);
    }
}
